### PR TITLE
feat: added std.GetHints for convenience. fixes #264

### LINKED ISSUE #264: hints in gadgets are not registered when packages are not imported
The problem appears when both the constraint system and witness are obtained by deserializing data.

Maybe add a new package which imports all hints in gadgets (`backend/hint/all` etc.) which has only a single side-effect of importing all hints from gadgets and registering. This package should then be imported by the compiler and provers.

### PR DESCRIPTION
**Pull request by gbotrel**
See #264. 

+ this PR adds a clear error message when a hint is missing for the solver; a compiled constraint system will be serialized with its hints dependencies; if any is missing at Solving time, the error message will contain the hint name.

### CODE REVIEW

File: internal/generator/backend/template/representations/r1cs.go.tmpl [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/consensys/gnark/frontend/compiled"
 	"github.com/consensys/gnark/frontend/schema"
 	"github.com/consensys/gnark/backend"
+	"github.com/consensys/gnark/backend"
```
Comment (ivokub) [inline internal/generator/backend/template/representations/r1cs.go.tmpl:15]: Duplicate entry.

File: internal/generator/backend/template/representations/r1cs.sparse.go.tmpl [hunk 2]
```diff
@@ -13,6 +13,7 @@ import (
 	"github.com/consensys/gnark/internal/backend/ioutils"
 	"github.com/consensys/gnark/frontend/compiled"
 	"github.com/consensys/gnark/backend"
+	"github.com/consensys/gnark/backend"
```
Comment (ivokub) [inline internal/generator/backend/template/representations/r1cs.sparse.go.tmpl:16]: Duplicate entry.

File: backend/backend.go [hunk 3]
```diff
@@ -72,6 +60,18 @@ type ProverConfig struct {
 	LoggerOut     io.Writer       // defaults to os.Stdout
 }
 
+// NewProverConfig returns a default ProverConfig with given prover options opts
+// applied.
+func NewProverConfig(opts ...ProverOption) (ProverConfig, error) {
+	opt := ProverConfig{LoggerOut: os.Stdout, HintFunctions: hint.GetAll()}
```
Comment (ivokub) [inline backend/backend.go:66]: Maybe also `std.GetHints()`?

**Review by ivokub — CHANGES_REQUESTED** [review]
Mostly looks good. I think the only point to consider if default prover config should include all hints from the gadget library or if there should be a prover option to include gadget hints.

File: backend/backend.go (hunk 3 above)
Comment (gbotrel) [reply backend/backend.go:66]: import cycle hell :) std uses test package to test itself.

Comment (ivokub) [reply backend/backend.go:66]: :(
In that case - I think it is good as is. It will probably require too much refactoring?

**Pull request merged**